distributed/utils_comm.py

```diff
@@ -275,3 +275,33 @@ def pack_data(o, d, key_types=object):
         return {k: pack_data(v, d, key_types=key_types) for k, v in o.items()}
     else:
         return o
+
+
+def subs_multiple(o, d):
+    """ Perform substitutions on a tasks
+
+    Parameters
+    ----------
+    o:
+        Core data structures containing literals and keys
+    d: dict
+        Mapping of keys to values
```

Comment:
With the current implementation these have to be *str* keys, may be worth noting.

Reply:
Hrmm that's a really good point. I think we'll want to cover generic, non-string keys too (thinking of, for example, a persisted dask array which has keys like `{("chunk", 0): <Future>}`).

Since `d` is a mapping which contains keys in the task graph to substitute, we could first check whether or not `o` is itself a key in `d`. That would let us know to make a substitution when we come across a key like, for example, `("chunk", 0)`. I pushed a commit with what I mean in code

Reply:
I *think* at this point all keys are strings already, but I may be wrong. I don't remember when keys are converted to strings (if it's on the client or the scheduler side), but at some point everything's a string, so the previous code may be fine. I was mostly commenting to update the docstring.

Reply:
The client converts keys to strings before sending to the scheduler.  The scheduler only understands string-valued keys.

Reply:
Thanks for pointing that out, I hadn't realized this string conversion took place. It looks like the the conversion happens here:

https://github.com/dask/distributed/blob/f7a0d7a8e5a729e84f695b8efb593e590c0ec2f4/distributed/client.py#L2446

a few lines after `Future`s have been inlined in the graph.

Reply:
Just updated `subs_multiple` with the improvements from @jcrist in https://github.com/dask/distributed/pull/3303#discussion_r355015487

Alternatively, we could try moving the `str_graph` call before substitutions take place. However, one advantage to the current `subs_multiple` implementation is we could use it elsewhere (on generic keys) should the need arise